### ISSUE TITLE
Usage from R

### ISSUE THREAD

**Issue by alimanfoo** [zarr-developers/community#18]
It would be great to be able to use zarr format data from R. This issue is intended for discussing options for enabling/supporting usage from R.

**Comment by gdkrmr** [comment]
I took a stab at wrapping z5 from R, it currently compiles, but it is not functional yet and there is still quite a lot of work to do, so don't judge me :-).
I am sharing this to avoid duplicated efforts, anyone who wants can join development

https://github.com/gdkrmr/zarr-R

**Comment by constantinpape** [comment]
> I took a stab at wrapping z5 from R, it currently compiles, but it is not functional yet and there is still quite a lot of work to do, so don't judge me

Let me know if you have any questions or need any help from the z5 site.

**Comment by alimanfoo** [comment]
Thanks a lot for sharing, great to hear about this!

**Comment by gdkrmr** [comment]
> Let me know if you have any questions or need any help from the z5 site.

What is the ETA for v2.0.0?

**Comment by constantinpape** [comment]
> Thanks, what is the ETA for v2.0.0?

The API redesign is done, I just need to test it a bit more.
Initially, my plan was to wait for the implementation of the S3 backend.
I have put in a bit of work into this, but it's not quite done yet and I don't have time to finish it right now.
(I was initially hoping for some external contributions to the S3 part, but this hasn't happened yet).

Anyway, I think I will release 2.0.0 without S3 or other cloud backends and push this to 2.1.0.
I can probably do it next week. I will let you know once it's there.

**Comment by jakirkham** [comment]
@gdkrmr, it would be great if you could hop on one of our meetings ( https://github.com/zarr-developers/community/issues/1 ), am sure others would be interested in hearing about your work and how we can help you.

**Comment by gdkrmr** [comment]
- I have something basic working working for z5 v2, currently only Int32 and Float64 (The builtin types of R).
- I still need to think how to deal with the other data types (e.g. Booleans).
- It needs a nice and simple high level interface, currently everything follows the z5 design pretty closely, which is not really suitable for the average R user.
- It produces an .so file that is almost 40Mb large :-)
- I will be happy to join your next meeting!

**Comment by constantinpape** [comment]
>     * I still need to think how to deal with the other data types (e.g. Booleans).

Just fyi, I don't support bool right now in z5.
@alimanfoo @jakirkham Are there any optimisations when zarr stores bools or is it storing a bool as one byte?

>     * It produces an .so file that is almost 40Mb large :-)

Interesting, for the python bindings the .so is quite a bit smaller,
~ 2.5 MB (build on Ubuntu 18 with gcc 7 and `Release`).

**Comment by jakirkham** [comment]
I don't think we are doing anything special. Though could imagine one implementing a bit packing codec.

Maybe there are some compiler flags that can help?

**Comment by constantinpape** [comment]
> Maybe there are some compiler flags that can help?

Probably yes.

@gdkrmr What operating system are you using and which compiler?
Are you using CMake? If so, maybe try compiling with `Release` or with `MinSizeRel`.

**Comment by alimanfoo** [comment]
On Mon, 27 Jan 2020, 21:05 Constantin Pape, <notifications@github.com>
wrote:

> * I still need to think how to deal with the other data types (e.g. Booleans).
>
> Just fyi, I don't support bool right now in z5.
> @alimanfoo <https://github.com/alimanfoo> @jakirkham
> <https://github.com/jakirkham> Are there any optimisations when zarr
> stores bools or is it storing a bool as one byte?
>
Same as numpy, bool as one byte.

**Comment by gdkrmr** [comment]
> > ```
> > * I still need to think how to deal with the other data types (e.g. Booleans).
> > ```
> 
> 
> Just fyi, I don't support bool right now in z5.
> 
> @alimanfoo @jakirkham Are there any optimisations when zarr stores bools or is it storing a bool as one byte?

R stores bools as bytes, because there is also a `NA` for bools. So I guess the way to go is to add an argument when reading to transform the data either into R integers or bools

> > Maybe there are some compiler flags that can help?
> 
> Probably yes.
> 
> @gdkrmr What operating system are you using and which compiler?
> Are you using CMake? If so, maybe try compiling with `Release` or with `MinSizeRel`.

Ubuntu 16.04 and I have to use the R build system, which uses Makefiles.

> > ```
> > * It produces an .so file that is almost 40Mb large :-)
> > ```
> 
> 
> Interesting, for the python bindings the .so is quite a bit smaller,
> 
> ~ 2.5 MB (build on Ubuntu 18 with gcc 7 and `Release`).

I can get the size of the `.so` down to < 1MB if I strip debug symbols or use link time optimization.
I have asked on the R developers mailing list and the CRAN (the official R package repository) policy is quite restrictive with these kinds of flags, so *they* have to live with it. Ironically their checker throws a warning if the `.so` gets too large :-). I just found that this was a curious fact, nothing to really worry about.

**Comment by gdkrmr** [comment]
What is the state of Zarr support in R? I haven't looked at my package for a while and wonder if someone else has done some work on this in the meanwhile or is planning to work on this?

**Comment by LTLA** [comment]
I'm late to the party, but Googling most permutations of "zarr for R" gives this thread as the top hit, then @gdkrmr's repo, and Bioconductor's **ZarrExperiment** (I'll get to this later). So I'd guess your stuff is still the best we've got right now.

If you're planning to keep working on your zarr R package, I'd be willing to test it out on some genome-scale data. I've been eyeing some alternatives to HDF5 for a while and would be very interested in building on top of whatever you make.

Our current approach in **ZarrExperiment** just does the simple thing of dispatching to the Python library via **reticulate**. A native port would be much preferred if it is feasible. If your package gets more mature, we would use it to create a `DelayedArray` backend for zarr that would work in all analysis pipelines as a plug-and-play replacement for HDF5.

(Maybe you should call the package **zarrr**, ho ho ho.)

**Comment by ocefpaf** [comment]
> I'm late to the party, but Googling most permutations of "zarr for R" gives this thread as the top hit, then @gdkrmr's repo, and Bioconductor's **ZarrExperiment** (I'll get to this later). So I'd guess your stuff is still the best we've got right now.

Same here. I'll be teaching a workshop for R users soon and I was wondering about zarr support. So far I got it via nczarr. See the last cell of [this notebook](https://nbviewer.jupyter.org/gist/ocefpaf/4a078b19db4fd5507d2d21691abaa689). But it would be nice to add alternatives that don't require a netcdf installation.

**Comment by joshmoore** [comment]
Would it help to get zarrrrrr interested parties together at the next [community meeting](https://github.com/zarr-developers/community/issues/1) (May 5th) to discuss a path forward?

From my side, I'd love to see one (or more?) R implementation in https://github.com/zarr-developers/zarr_implementations/

cc: @gdkrmr @keller-mark @ocefpaf @LTLA (@dominikl? @jkh1?)

**Comment by jkh1** [comment]
Count me in. As a regular R user, this is something I've been thinking about recently. I'd favour the C++/Rcpp path over the reticulate approach as I've had issues with reticulate before (in my experience, R doesn't always play well with the various python envs/conda).

**Comment by davidbrochart** [comment]
Maybe we could provide R bindings of xtensor-zarr? We already do that [for xtensor](https://github.com/xtensor-stack/xtensor-r), and there exists [an R package for xtensor](https://github.com/xtensor-stack/Xtensor.R) already. We could improve this package so that it allows Zarr access, and users could use the same package for array processing. The package would then be equivalent to something like Zarr + NumPy.

**Comment by gdkrmr** [comment]
the `netcdf-c` library has added support for zarr files. `netcdf-c` is the basis for the R package `ncdf4`. There are discussion on how to get it working in R https://github.com/Unidata/netcdf-c/issues/1982.

**Comment by keller-mark** [comment]
This sounds great! I started an extremely rough pure R function for producing a single Zarr chunk from an R matrix here https://github.com/vitessce/vitessce-r/blob/keller-mark/zarr/R/zarr.R#L215 in case anyone is interested. Unfortunately I cannot attend at 2pm eastern time on May 5th due to a conflict but perhaps @ilan-gold @manzt @th789 @mccalluc are  interested

**Comment by gdkrmr** [comment]
I will try to attend but cannot make any promises.

**Comment by joshmoore** [comment]
Looks like the time slot didn't work out for R folks. No worries. Note that the [19th](https://github.com/zarr-developers/community/issues/1#issuecomment-832935477) is cancelled; we'll be back on the regular zoom on the 2nd though. If a different time slot would be better, feel free to say the word.

**Comment by ocefpaf** [comment]
> the `netcdf-c` library has added support for zarr files. `netcdf-c` is the basis for the R package `ncdf4`. There are discussion on how to get it working in R [Unidata/netcdf-c#1982](https://github.com/Unidata/netcdf-c/issues/1982).

In a way that already works. See the last cell of https://nbviewer.jupyter.org/gist/ocefpaf/4a078b19db4fd5507d2d21691abaa689

But nczarr is not exactly the same as zarr. I'm not well versed in the details but maybe a core zarr (c/c++/rust, whatever) that we can wrap in Python and R is still needed?

**Comment by joshmoore** [comment]
@ocefpaf : I only know what's on the [docs](https://www.unidata.ucar.edu/blogs/developer/entry/overview-of-zarr-support-in) and what I've tested on the CLI, but my understanding was that nczarr has a mode to work with pure Zarr that may be of interest. I'd defer to @DennisHeimbigner whether a portion of the library could be used as a core.

**Comment by DennisHeimbigner** [comment]
Josh is correct. We support pure zarr read/write, so as long as you are willing to live
with the restricted meta-data of pure zarr you can use netcdf-c for pure V2 zarr.
The next netcdf-c release (version 4.8.1) will also add support for the Xarray convention
for named dimensions.
As for pulling out pieces, that is doable. As is usual, the documentation could be improved.
Much of the code uses the netcdf internal data structures for implementing the netcdf-C
API. But at least these parts might be usable.
1. the code caching and read/write of chunks
2. the code that reads/writes zarr metadata.
3. the code that wraps access to the underlying storage e.g. files, zip file, and S3.

**Comment by DennisHeimbigner** [comment]
BTW you could try this experiment with R wrapping netcdf-4.8.0
1. Take a simple R program that creates a netcdf .nc file, call it simple.nc
2. Modify the program so that instead of calling whatever the R equivalent of nc_create("simple.nc"...) instead call the equivalent of nc_create("file://simple.zarr#mode=zarr,file",...)

This should create a directory called "simple.zarr" that contains a pure zarr container.
The name "simple.zarr" is not special, you can call it whatever you want.
If you try this then let me know what happens.

**Comment by jkh1** [comment]
[Related discussion on image.sc](https://forum.image.sc/t/support-for-zarr-in-r).

**Comment by schienstockd** [comment]
I have a beginner's question to opening zarr files with `netcdf`.

I have built `netcdf-c` with zarr support (I think) and then built the R package `ncdf4`.
I am not sure how to open the dataset up then. I have an OME-ZARR file generated from `biofromats2raw` and tried to open the file like this:

```R
library(ncdf4)

# open file
ncin <- nc_open(
  "file:///Users/me/image.ome.zarr#mode=nczarr,zarr"
)
ncin
# Error in R_nc4_inq: NetCDF: Invalid argument
# Error in nc_get_grp_info(gids[ib], root_group$fqgn, format) : 
#   nc_get_grp_info: R_nc4_inq returned error on group id 524289
```

**Comment by gdkrmr** [comment]
I couldn't get it to work either, see https://github.com/Unidata/netcdf-c/issues/1982

**Comment by DennisHeimbigner** [comment]
A couple of things.
1. It appears you are using netcdf-c version 4.8.1 correct?
2. try this command to avoid any R interference.
     ncdump -h  "file:///Users/me/image.ome.zarr#mode=nczarr,zarr"

**Comment by DennisHeimbigner** [comment]
BTW what operating system are you using?

**Comment by schienstockd** [comment]
`MacOS Catalina 10.15.5`

It is a a multiscale OME-zarr where the image is in the path `/0/0` of the file.
When I try to read from the root then the file is not found.
When I try to read from the dataset path directly the header seems to be empty:

```bash
> ncdump -h "file:///Users/me/ccidImage.ome.zarr#mode=nczarr,zarr"
ncdump: file:///Users/me/ccidImage.ome.zarr#mode=nczarr,zarr: No such file or directory

> ncdump -h "file:///Users/me/ccidImage.ome.zarr/0/0#mode=nczarr,zarr"
netcdf \0 {
}
```

```bash
nc-config --all

This netCDF 4.8.1-development has been built with the following features: 

  --cc            -> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
  --cflags        -> -I/usr/local/include
  --libs          -> -L/usr/local/lib -lnetcdf
  --static        -> -lhdf5_hl -lhdf5 -lsz -lz -ldl -lm -lsz -lcurl -lzip

  --has-c++       -> no
  --cxx           -> 

  --has-c++4      -> yes
  --cxx4          -> /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++
  --cxx4flags     -> -I/usr/local/Cellar/osgeo-netcdf/4.7.4/include
  --cxx4libs      -> -L/usr/local/Cellar/osgeo-netcdf/4.7.4/lib -lnetcdf-cxx4 -lnetcdf

  --has-fortran   -> yes
  --fc            -> /usr/local/bin/gfortran
  --fflags        -> /usr/local/Cellar/osgeo-netcdf/4.7.4/include
  --flibs         -> -L/usr/local/Cellar/osgeo-netcdf/4.7.4/lib
  --has-f90       -> TRUE
  --has-f03       -> FALSE

  --has-dap       -> yes
  --has-dap2      -> yes
  --has-dap4      -> yes
  --has-nc2       -> yes
  --has-nc4       -> yes
  --has-hdf5      -> yes
  --has-hdf4      -> no
  --has-logging   -> no
  --has-pnetcdf   -> no
  --has-szlib     -> yes
  --has-cdf5      -> yes
  --has-parallel4 -> no
  --has-parallel  -> no
  --has-nczarr    -> yes

  --prefix        -> /usr/local
  --includedir    -> /usr/local/include
  --libdir        -> /usr/local/lib
  --version       -> netCDF 4.8.1-development
```

**Comment by joshmoore** [comment]
> When I try to read from the dataset path directly the header seems to be empty:

Could this be related to the "dimension_separator" metadata, @DennisHeimbigner ? @schienstockd , can you show us the content of `0/0/.zarray`?

**Comment by DennisHeimbigner** [comment]
I think I see the problem. I use a heuristic to break a key into the variable key
and the chunk index/key. The heuristic says to get the longest suffix of integers
as the chunk index. So, in this case it is eating up too much of the key as the chunk index.
I can fix, but out of curiosity why do you have a  variable named "0"

**Comment by schienstockd** [comment]
@joshmoore 

`0/0/.zarray`
```json
{
  "chunks" : [
    1,
    1,
    1,
    512,
    512
  ],
  "compressor" : {
    "clevel" : 5,
    "blocksize" : 0,
    "shuffle" : 1,
    "cname" : "lz4",
    "id" : "blosc"
  },
  "dtype" : ">u2",
  "fill_value" : 0,
  "filters" : null,
  "order" : "C",
  "shape" : [
    180,
    4,
    8,
    512,
    512
  ],
  "zarr_format" : 2,
  "dimension_separator" : "/"
}
```

I am not sure where the '0' variable comes from .. I used `bioformats2raw` to convert the image

**Comment by jakirkham** [comment]
Looks like there is a very rough Zarr implementation in R

https://github.com/keller-mark/pizzarr

cc @keller-mark (hopefully I've clarified that correctly; please feel free to correct me if not)

**Comment by keller-mark** [comment]
Yes very rough indeed. Of course open to contributions or more detailed feature requests / issues.

**Comment by joshmoore** [comment]
See discussion post under https://github.com/zarr-developers/zarr-python/discussions/1088

cc: @mike-lawrence

**Comment by bart1** [comment]
The `stars` package seems to have a implementation (I did not test it): https://r-spatial.org/r/2022/09/13/zarr.html

**Comment by mike-lawrence** [comment]
> The `stars` package seems to have a implementation (I did not test it): https://r-spatial.org/r/2022/09/13/zarr.html

I think stars only provides read access, no write.

**Comment by mike-lawrence** [comment]
Seems to be solid progress [here](https://github.com/vitessce/vitessceR/blob/main/R/data_to_zarr.R)

**Comment by jkh1** [comment]
The [Rarr package](https://bioconductor.org/packages/release/bioc/html/Rarr.html) is now on Bioconductor. The [repository is here](https://github.com/grimbough/Rarr). It's written in C and writing is supported  although for now limited to double and string types.

**Comment by mike-lawrence** [comment]
> The [Rarr package](https://bioconductor.org/packages/release/bioc/html/Rarr.html) is now on Bioconductor. The [repository is here](https://github.com/grimbough/Rarr). It's written in C and writing is supported  although for now limited to double and string types.

Cool! I always forget to check bioconductr for packages 🤦‍♂️

**Comment by keller-mark** [comment]
Hi all, update on [pizzarr](https://github.com/keller-mark/pizzarr): some things are working now!

- Reading/writing of integer and float arrays (v2 spec)
- 3 stores: MemoryStore, DirectoryStore, and HttpStore.
- 2 types of compression: LZ4 and Zstd
- Convenience functions for arrays/groups
- R-like (one-based) and Python-like (zero-based) slicing
- List of some remaining features to implement at https://github.com/keller-mark/pizzarr/issues

I have updated the docs a bit, with a simple OME-NGFF demo at https://keller-mark.github.io/pizzarr/articles/ome-ngff.html

<img width="1605" alt="Screenshot 2023-08-19 at 6 06 44 PM" src="https://github.com/zarr-developers/community/assets/7525285/1ff9db1b-35eb-46ee-bc33-f011c0748964">

**Comment by sanketverma1704** [comment]
Thanks for working on Pizzarr and updating us, @keller-mark.

May I add this to our website (https://zarr.dev/implementations/)?

**Comment by keller-mark** [comment]
@MSanKeys963 Yes feel free to add! Thanks!

**Comment by mdsumner** [comment]
I'm working on multidim GDAL -which can read and write Zarr - in this branch of gdalraster (I'm not the core maintainer). It's very bare now but I'm encouraged by progress so far. Rcpp modules is a great way to expose libraries.

https://github.com/mdsumner/gdalraster/tree/multidimnew

All  multidim code is in src/gdalmultidimraster.h/.cpp

I'm interested in pizzarr too, hoping to get across how it works and contribute. Neither GDAL or pizzarr support virtual references though, which I think is a serious problem for R. I've discussed prospects for this in GDAL and there's wider interest but no plans yet.

**Comment by dblodgett-usgs** [comment]
Hi All -- 

Exciting news. `pizzarr` 🍕 is now on zarr-developers: https://github.com/zarr-developers/pizzarr I don't have privileges to modify setting yet, so I can't get the github pages updated, but I am working actively to get it working with Zarr V3 and up on CRAN. If anyone is interested/willing to help with some review and feedback. It would be very welcome!

**Comment by dblodgett-usgs** [comment]
Update from me -- pizzarr is now on CRAN. 

[Slice into Zarr Arrays in R • pizzarr](https://zarr.dev/pizzarr/) 

Please test and submit issues or PRs as you find things that could be better.
